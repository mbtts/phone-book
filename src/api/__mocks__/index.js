const mockData = [
  {
    id: 0,
    name: "Oleta Level",
    phone_number: "+442032960159",
    address: "10 London Wall, London EC2M 6SA, UK"
  },
  {
    id: 6,
    name: "Stanley Vanderhoof",
    phone_number: "+442032960000",
    address: "17 Anchor Ave, Darwen BB3 0AZ, UK"
  }
];

const api = jest.fn().mockImplementation(() => {
  return Promise.resolve(mockData);
});

export { api as default, mockData };
