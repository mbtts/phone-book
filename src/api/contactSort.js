const createLastName = contact => {
  const lastName = contact.name.match(/\S+/g).pop();
  return lastName;
};

const sortByLastName = modifier => {
  return (a, b) =>
    createLastName(a).localeCompare(createLastName(b)) * modifier;
};

const sortByLastNameAsc = sortByLastName(1);
const sortByLastNameDesc = sortByLastName(-1);

export { sortByLastNameAsc, sortByLastNameDesc };
