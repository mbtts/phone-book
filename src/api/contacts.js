const api = async () => {
  const result = await fetch(
    "https://www.mocky.io/v2/581335f71000004204abaf83"
  );
  const json = await result.json();
  const indexed = json.contacts.map((contact, i) => ({ id: i, ...contact }));
  return indexed.sort(sortByLastNameAsc);
};

const createLastName = contact => {
  const lastName = contact.name.match(/\S+/g).pop();
  return lastName;
};

const sortByLastName = modifier => {
  return (a, b) => {
    return createLastName(a).localeCompare(createLastName(b)) * modifier;
  };
};

const sortByLastNameAsc = sortByLastName(1);
const sortByLastNameDesc = sortByLastName(-1);

export { api as default, sortByLastNameAsc, sortByLastNameDesc };
