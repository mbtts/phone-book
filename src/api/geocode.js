const GLOBAL_INIT = "initMaps";

const cache = {};

const processResults = (address, results, resolve, reject) => {
  if (results[0]) {
    const location = results[0].geometry.location;
    const coords = { lat: location.lat(), lng: location.lng() };
    cache[address] = coords;
    resolve(coords);
  } else {
    reject("No results found");
  }
};

const geoCode = address => {
  return new Promise((resolve, reject) => {
    if (cache[address]) {
      return resolve(cache[address]);
    }

    const geocoder = new google.maps.Geocoder();
    geocoder.geocode({ address }, (results, status) => {
      status === "OK"
        ? processResults(address, results, resolve, reject)
        : reject(`Geocode error ${status}`);
    });
  });
};

const createInitialiser = resolve => () => {
  window[GLOBAL_INIT] = undefined;
  resolve(google);
};

const initialiseMapsApi = () => {
  return new Promise(resolve => {
    if (typeof google !== "undefined") {
      return resolve(google);
    }

    window[GLOBAL_INIT] = createInitialiser(resolve);
    const script = document.createElement("script");
    script.setAttribute(
      "src",
      `https://maps.googleapis.com/maps/api/js?key=${
        process.env.GOOGLE_MAPS_API
      }&callback=${GLOBAL_INIT}`
    );
    document.body.appendChild(script);
  });
};

export { geoCode, initialiseMapsApi };
