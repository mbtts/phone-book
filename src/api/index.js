const api = async () => {
  const result = await fetch("http://www.mocky.io/v2/581335f71000004204abaf83");
  const json = await result.json();
  return json.contacts;
};

export default api;
