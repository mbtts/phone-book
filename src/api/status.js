const status = {
  LOADING: "LOADING",
  LOADED: "LOADED",
  ERROR: "ERROR"
};

export default status;
