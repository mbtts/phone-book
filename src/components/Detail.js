import { ERROR, LOADED, LOADING } from "../api/status";
import React, { Fragment, PureComponent } from "react";

import PropTypes from "prop-types";

class Detail extends PureComponent {
  constructor(props) {
    super(props);
  }

  render() {
    const { status, contact } = this.props;

    let view;
    if (status === LOADING) {
      view = <p>Loadingâ€¦</p>;
    } else if (status === ERROR) {
      view = <p>Error</p>;
    } else if (status === LOADED && contact) {
      view = (
        <Fragment>
          <h3>{contact.name}</h3>
          <p>
            <a href={`tel:${contact.phone_number}`}>{contact.phone_number}</a>
          </p>
          <p>{contact.address}</p>
        </Fragment>
      );
    }

    return view;
  }
}

Detail.propTypes = {
  contact: PropTypes.shape({
    name: PropTypes.string.isRequired,
    phone_number: PropTypes.string.isRequired,
    address: PropTypes.string.isRequired
  })
};

export default Detail;
