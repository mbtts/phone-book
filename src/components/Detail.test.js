import { ERROR, LOADED, LOADING } from "../api/status";

import Detail from "./Detail";
import React from "react";

const mockData = {
  id: 6,
  name: "Stanley Vanderhoof",
  phone_number: "+442032960000",
  address: "17 Anchor Ave, Darwen BB3 0AZ, UK"
};

describe("Detail view component", () => {
  it("should render initial state", () => {
    const wrapper = shallow(<Detail status={LOADED} contact={mockData} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should be loading initially", () => {
    const wrapper = shallow(<Detail status={ERROR} contact={null} />);
    expect(wrapper.find("p").text()).toEqual("Error");
  });

  it("should display an error if the detail cannot load", () => {
    const wrapper = shallow(<Detail status={LOADING} contact={null} />);
    expect(wrapper.find("p").text()).toEqual("Loadingâ€¦");
  });
});
