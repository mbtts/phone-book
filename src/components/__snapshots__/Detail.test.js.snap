// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail view component should render initial state 1`] = `
<React.Fragment>
  <h3>
    Stanley Vanderhoof
  </h3>
  <p>
    <a
      href="tel:+442032960000"
    >
      +442032960000
    </a>
  </p>
  <p>
    17 Anchor Ave, Darwen BB3 0AZ, UK
  </p>
</React.Fragment>
`;
